core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -691,48 +692,67 @@ class LogManager(logDirs: Seq[File],
           else
             nextLogDir().getAbsolutePath
         }
-        if (!isLogDirOnline(logDir))
-          throw new KafkaStorageException(s"Can not create log for $topicPartition because log directory $logDir is offline")
-
-        try {
-          val dir = {
-            if (isFuture)
-              new File(logDir, Log.logFutureDirName(topicPartition))
-            else
-              new File(logDir, Log.logDirName(topicPartition))
-          }
-          Files.createDirectories(dir.toPath)
-
-          val log = Log(
-            dir = dir,
-            config = config,
-            logStartOffset = 0L,
-            recoveryPoint = 0L,
-            maxProducerIdExpirationMs = maxPidExpirationMs,
-            producerIdExpirationCheckIntervalMs = LogManager.ProducerIdExpirationCheckIntervalMs,
-            scheduler = scheduler,
-            time = time,
-            brokerTopicStats = brokerTopicStats,
-            logDirFailureChannel = logDirFailureChannel)
 
+        val logDirName = {
           if (isFuture)
-            futureLogs.put(topicPartition, log)
+            Log.logFutureDirName(topicPartition)
           else
-            currentLogs.put(topicPartition, log)
+            Log.logDirName(topicPartition)
+        }
 
-          info(s"Created log for partition $topicPartition in $logDir with properties " +
-            s"{${config.originals.asScala.mkString(", ")}}.")
-          // Remove the preferred log dir since it has already been satisfied
-          preferredLogDirs.remove(topicPartition)
+        val dir = createLogDirectory(logDir, logDirName)
+          .recoverWith {
+            case e: KafkaStorageException =>
+              // If the next log directory, try the other ones before failing
+              _liveLogDirs.asScala
```

Comment:
I have a minor suggestion. This should be a rare case, so it may not matter too much, but could we let `nextLogDir` return the full ordered list of suggested directories rather than just a single entry? Then we would still take load into account when retrying. It would also let us move this loop to the top level so that `createLogDirectory` is only called in spot.

Something roughly like this:
```scala
val logDirs = {

  // omitted

  if (preferredLogDir != null)
    List(preferredLogDir)
  else
    nextLogDirs()
}

val dir = logDirs
  .find(createLogDirectory(_, logDirName).isSuccess)
  .get
```

Reply:
Thanks for the review @hachikuji. I like the suggestion, makes things simpler. Added this in b6c36f1. Please take another look.